Update special handling of autoconf

sbc100 opened the pull request:
- Avoid running an inner emcc process.

- Separate out function for adding `#!` line.  We could add a user
  option for this if there is demand.

- Allow `#!` line to be configured via new JS_ENGINE setting.
  This is different from the plural JS_ENGINES used by test
  code since it never makes sense to have more than one.
  It also makes it easier to set vai the command line. For
  example to build something that is runnable under d8 you
  can just do:
    `EM_JS_ENGINE=d8 emcc hello.c -o out.js -sEXECUTABLE`
  This setting defaults to JS_ENGINES[0] which itself defaults
  to NODE_JS so it should be very rare that anyone actually
  needs to set this.

sbc100 commented:
@surma

surma commented:
Would this also allow me to do `EM_JS_ENGINE="node --experimental-wasm-simd" emcc hello.c -o out.js -sEXECUTABLE`?

sbc100 commented:
> Would this also allow me to do `EM_JS_ENGINE="/path/to/node --experimental-wasm-simd" emcc hello.c -o out.js -sEXECUTABLE`?

In my initial version I wanted to allow that by using shlex.split on the strings.    But in the general case this wouldn't for strings coming from the config file since it would break things like "NODE_JS = "/path with spaces/bin/node"'.   So I decided not to do that as part of this PR... but I will look into making that work again now, maybe just for environment variables.

sbc100 commented:
Done!

kripken commented on emcc.py:
```diff
@@ -1467,7 +1432,7 @@ def check(input_file):
       cflags.append('-DEMSCRIPTEN')
 
     # Treat the empty extension as an executable, to handle the commond case of `emcc -o foo foo.c`
```
please add a comment here

or maybe `('.out', '')` could be another constant, `OTHER_EXECUTABLE_ENDINGS`?

kripken commented on src/settings.js:
```diff
@@ -1799,6 +1799,11 @@ var WASM2C = 0;
+// Add a #! line output JavaScript file and make it executable.  This command
+// used to run the output can be configured via the .emscripten config file using
+// the JS_ENGINE setting and defaults to NODE_JS.
+var EXECUTABLE = '';
```
how about `EXECUTABLE_JS`? or do you think it'll be useful for non-JS too in standalone mode or elsewhere?

sbc100 replied:
I don't really care what this is called.  If you prefer that name I will change it.

I could even make this an internal setting in this change?  But it makes it less testable I guess?

kripken replied:
I'm ok with `EXECUTABLE`, just wanted to raise some questions.

kripken commented on src/settings.js:
```diff
@@ -1799,6 +1799,11 @@ var WASM2C = 0;
 // in a custom location.
 var SEPARATE_DWARF_URL = '';
 
+// Add a #! line output JavaScript file and make it executable.  This command
```
```suggestion
// Add a #! line to the JavaScript output file and make it executable.  The command
```

kripken commented on tools/shared.py:
```diff
@@ -367,11 +369,17 @@ def parse_config_file():
 
+  # A subset of the above keys that represent commands. In this case its useful to
+  # use shlex.split on the environment variable.
+  COMMAND_KEYS = ('NODE_JS', 'JAVA', 'WASMER', 'WASMTIME', 'JS_ENGINE', 'CLOSURE_COMPILER')
```
this change seems slightly risky and maybe might be worth splitting out?

sbc100 replied:
Use you are right there is a slight risk.  The user who would could be broken by this would be somebody who does `EM_NODE_JS='/path/with spaces/bin/node'` because now that would be interpreted as two different args.

I can split this out and add a test for it.   I don't think it should be too risky since I don't imagine there are many users of EM_XXX environment variables who also want to include pathnames with spaces (normally thats mostly windows users).

kripken commented on emcc.py:
```diff
@@ -1246,6 +1199,20 @@ def get_last_setting_change(setting):
+      # Add `#!` line to output JS and make it executable.
+      shared.Settings.EXECUTABLE = 1
+      # Autoconf expects the executable output file to be called `a.out`
+      default_target_name = 'a.out'
```
another option, which I'm not sure would be better, is to not do this here, and not make the other changes for `.out` suffix support, and instead if `EXECUTABLE` is set to do something at the very end of compilation (rename or copy `a.out.js` to `a.out`).

sbc100 replied:
I think I prefer this way rather than doing more copying/renaming.

I might even one day see if we can change the default globally to match clang/gcc default output name .. but that is much bigger change :)

sbc100 commented:
I removed the new setting, and removed the new shlex features for parsing env vars.

The former we can look at adding if anyone actually wants to use it.  The later I will upload as a separate PR. 

I think there is also a followup to look at removing JS_CONTAINING_ENDINGS but this PR should be more precise now.
